crates/ruff_python_formatter/src/comments/placement.rs

```diff
@@ -633,6 +651,123 @@ fn handle_parameters_separator_comment<'a>(
+/// and a subsequent body.
+///
+/// For example, given:
+/// ```python
+/// if True:
+///     pass
+/// ```
+///
+/// `True` would be the `expr`, and `pass` would be the only statement in the `body`.
+///
+/// Or, given:
+/// ```python
+/// for x in range(10):
+///     pass
+/// ```
+///
+/// `x in range(10)` would be the `expr`, and `pass` would be the only statement in the `body`.
+#[derive(Debug)]
+struct TerminalExpression<'a> {
+    /// The test or iterable expression.
+    expr: &'a Expr,
+    /// The body of the statement.
+    body: &'a [Stmt],
+}
+
+impl<'a> From<&'a ast::StmtIf> for TerminalExpression<'a> {
+    fn from(stmt: &'a ast::StmtIf) -> Self {
+        Self {
+            expr: &stmt.test,
+            body: &stmt.body,
+        }
+    }
+}
+
+impl<'a> From<&'a ast::StmtWhile> for TerminalExpression<'a> {
+    fn from(stmt: &'a ast::StmtWhile) -> Self {
+        Self {
+            expr: &stmt.test,
+            body: &stmt.body,
+        }
+    }
+}
+
+impl<'a> From<&'a ast::StmtFor> for TerminalExpression<'a> {
+    fn from(stmt: &'a ast::StmtFor) -> Self {
+        Self {
+            expr: &stmt.iter,
+            body: &stmt.body,
+        }
+    }
+}
+
+impl<'a> From<&'a ast::StmtAsyncFor> for TerminalExpression<'a> {
+    fn from(stmt: &'a ast::StmtAsyncFor) -> Self {
+        Self {
+            expr: &stmt.iter,
+            body: &stmt.body,
+        }
+    }
+}
+
+impl<'a> From<&'a ast::StmtWith> for TerminalExpression<'a> {
+    fn from(stmt: &'a ast::StmtWith) -> Self {
+        let item = stmt.items.last().unwrap();
+        Self {
+            expr: item.optional_vars.as_deref().unwrap_or(&item.context_expr),
+            body: &stmt.body,
+        }
+    }
+}
+
+impl<'a> From<&'a ast::StmtAsyncWith> for TerminalExpression<'a> {
+    fn from(stmt: &'a ast::StmtAsyncWith) -> Self {
+        let item = stmt.items.last().unwrap();
+        Self {
+            expr: item.optional_vars.as_deref().unwrap_or(&item.context_expr),
+            body: &stmt.body,
+        }
+    }
+}
+
+/// Mark `# comment` as a trailing comment on the test (`True`) in:
+/// ```python
+/// while (
+///     True
+///     # comment
+/// ):
+///     pass
+/// ```
+fn handle_terminal_comment<'a>(
+    comment: DecoratedComment<'a>,
+    terminal: &TerminalExpression<'a>,
+    locator: &Locator,
+) -> CommentPlacement<'a> {
+    // The comment needs to be positioned after the expression, but before the body...
+    if comment.start() > terminal.expr.end()
+        && terminal
+            .body
+            .first()
+            .is_some_and(|first| first.start() > comment.end())
+    {
+        // It _also_ needs to be positioned before the colon...
+        if let Some(colon) = SimpleTokenizer::starts_at(terminal.expr.end(), locator.contents())
+            .skip_trivia()
+            .skip_while(|token| token.kind == SimpleTokenKind::RParen)
+            .find(|token| token.kind == SimpleTokenKind::Colon)
```

Comment:
isn't that handled by making sure we're after the check?

Reply:
It probably works fine, but `next` is the proper method here to make sure this returns false if we made any wrong assumptions. It also is faster, because `find` performs a loop, which we don't need. 

The alternative is to remove the `skip_while` and just use the `find`. Which is probably fine, assuming we correctly considered all edge cases.